### PR TITLE
[contracts] feat: create proposal contract

### LINKED ISSUE #23: Smart Contract: ProposalContract
## Smart Contract: ProposalContract

**Difficulty:** Advanced  
**Estimated Time:** 10–14 hours  

### Goal
Develop the core **ProposalContract** for Harmonia. This Soroban smart contract will handle the creation, voting, and lifecycle management of DAO proposals.

### Responsibilities
- Store proposals on-chain with metadata
- Allow authorized members to vote
- Track vote counts (For, Against, Abstain)
- Enforce deadlines and proposal status changes (open/closed/accepted/rejected)
- Emit events for frontend tracking

### Required Functions
```rust
fn create_proposal(
  env: Env,
  title: Symbol,
  description: Symbol,
  deadline: u64,
  proposal_type: Symbol
)

fn vote(
  env: Env,
  proposal_id: u32,
  vote_choice: Symbol // For, Against, Abstain
)

fn finalize(
  env: Env,
  proposal_id: u32
)

fn get_votes(
  env: Env,
  proposal_id: u32
) -> (u32, u32, u32)

fn get_proposal(
  env: Env,
  proposal_id: u32
) -> Proposal
```

### Proposal Structure Example
```rust
struct Proposal {
  id: u32,
  title: Symbol,
  description: Symbol,
  created_at: u64,
  deadline: u64,
  proposal_type: Symbol,
  status: Symbol,
  for_votes: u32,
  against_votes: u32,
  abstain_votes: u32,
  created_by: Address
}
```

### Events
- `proposal_created`
- `vote_cast`
- `proposal_finalized`

### Access Control
- Use `require_auth(env.invoker())`
- Restrict voting to valid members (can integrate later with MemberRegistry)

### Storage Keys
- `proposal:<id>` → Proposal
- `vote:<proposal_id>:<address>` → vote_choice

### Testing Requirements
- Unit tests for:
  - Creating proposals
  - Casting votes
  - Preventing double voting
  - Auto-finalization logic

### Acceptance Criteria
- [ ] Contract compiles with `soroban build`
- [ ] Tests run with `soroban test`
- [ ] Schema and metadata generate correctly
- [ ] Event logs are emitted on all main actions

### Bonus
- Add support for proposal categories (treasury, governance, system)
- Add optional quorum enforcement

### How to Claim
1. Comment below to claim this issue  
2. Clone the repo and work inside `/contracts/proposal/`  
3. Use `soroban build` and `soroban test` to validate  
4. Open a PR referencing this issue with the contract + tests

This is a foundational contract for Harmonia's governance and will power the entire /proposals dashboard.

### PR DESCRIPTION
**Pull request by derianrddev**
## 🛠️ Issue  
- Closes #23 

## 📖 Description  
Implements the core `ProposalContract` for Harmonia, enabling DAO proposal creation, voting, finalization, and on-chain storage with event emissions.

## ✅ Changes made  
- Added `create_proposal`, `vote`, `finalize`, `get_votes`, and `get_proposal` functions  
- Stored proposals and votes on-chain with typed keys  
- Emitted `proposal_created`, `vote_cast`, and `proposal_finalized` events  
- Enforced deadlines, quorum, and status transitions  
- Included full unit test coverage for all main flows and edge cases

## 🖼️ Media (screenshots/videos)  

image

image

### CODE REVIEW

File: apps/contracts/Cargo.toml [hunk 1]
```diff
@@ -2,7 +2,8 @@
 resolver = "2"
 members = [
   "role-manager",
-  "referral-contract"
+  "referral-contract", 
+  "proposal"
```
Comment (aguilar1x) [inline apps/contracts/Cargo.toml:6]: change the name contract to "proposal-contract"

**Review by aguilar1x — CHANGES_REQUESTED** [review]
Add contract errors to handle them properly.

**Review by aguilar1x — APPROVED** [review]
LGTM!

**Review by DiegoB1911 — APPROVED** [review]

**Review by salazarsebas — APPROVED** [review]
LGTM!
